### ISSUE TITLE
Short-text classification approaches for title + URL

### ISSUE THREAD

**Issue by ChmaraX** [ChmaraX/forensix#136]
## Question

What are the approaches for classifying a browsing-history entry from **page title + URL together**, and which candidates survive v2's constraints?

[URL classifier replacement options](https://github.com/ChmaraX/forensix/issues/119) answered a different question and its conclusion does not carry. It surveyed **URL-only** classification, which is genuinely hard — that difficulty is why a list-first design looked competitive. But Chrome's `urls` table carries a title:

```
CREATE TABLE urls (id INTEGER PRIMARY KEY AUTOINCREMENT,
  url LONGVARCHAR, title LONGVARCHAR, visit_count INTEGER DEFAULT 0 NOT NULL,
  typed_count INTEGER DEFAULT 0 NOT NULL, last_visit_time INTEGER NOT NULL,
  hidden INTEGER DEFAULT 0 NOT NULL)
```

`components/history/core/browser/url_database.cc`, `URLDatabase::CreateURLTable` — confirmed against Chromium `main`.

`https://www.reddit.com/r/investing/comments/1a2b3c/` is nearly signal-free. `Best index funds for 2026 : r/investing` is not. This is ordinary short-text classification, and the approach space must be re-surveyed on that basis.

## Survey the approach space first, not model names

Characterise each family before naming candidates:

1. Single-pass discriminative classifier (fine-tuned encoder + head)
2. Multi-pass entailment (zero-shot NLI cross-encoder)
3. Single-pass embedding + label or exemplar similarity
4. Generative (small local LLM, constrained decoding)
5. Deterministic rules / domain lists

For each: **forward passes per row** as a function of taxonomy size, what supervision it needs, what it emits when uncertain, and how a decision is explained in a forensic report.

The cost model is the part that decides this. A 14-label taxonomy costs 14x more in an entailment design than a single-pass one, and that is an architectural fact available before any benchmark.

## Enumerate candidates systematically

Record the exact query used for each sweep so coverage is auditable:

- MTEB leaderboard, filtered to permissive licence / ONNX available / <150MB quantised
- HF API sweeps across `zero-shot-classification`, `text-classification`, `feature-extraction`, each intersected with `onnx`
- Published work on short-text and webpage classification, to catch families the HF taxonomy does not surface

**Trap to avoid:** an earlier informal pass picked candidates by keyword search and then benchmarked whatever it happened to find. Five encoders were tested because they were already known, not because they ranked. A shortlist assembled that way cannot support "best" — only "best of what I remembered".

## Hard filters — reject with the reason logged

Fully offline · deterministic across runs and machines · no Python at runtime · no `trust_remote_code` · no pickle deserialisation · weights redistributable under a stated licence

Licence must come from the model card or LICENSE file of the **authoritative** repo, not a mirror's tag. Two traps already found in #119 stand: UT1 is CC BY-SA 4.0 copyleft, and Tranco's default list embeds a CC BY-NC source. A third: ONNX mirrors frequently declare no licence at all even when the upstream is MIT.

## Taxonomy

Propose the v2 category set and justify each label as a forensic question someone actually asks. v1 used DMOZ top level (`Business, Arts, Computers, Reference, Society, Kids, Health, Science`) — a web-directory taxonomy with no Finance, Gambling, Adult, or Search. Note where candidate models impose a conflicting label set, since that is a hard constraint on which candidates are usable at all.

## Deliverable

`docs/research/136-approach-survey.md`

Every claim tagged CONFIRMED (with URL) or STILL-UNKNOWN. Sizes from real file listings, never parameter-count estimates. **No benchmark numbers in this ticket** — approaches, candidates, and constraints only. Measurement belongs to the harness ticket.

## Done when

A ranked shortlist of at most 6 candidates spanning at least 3 approach families, each surviving the hard filters, with the enumeration method recorded well enough that someone else could reproduce the shortlist.

**Comment by ChmaraX** [comment]
Delivered on `research/wayfinder-v2`:

- **[research/136-approach-survey.md](https://github.com/ChmaraX/forensix/blob/63c88f2/research/136-approach-survey.md)** — the survey (note: repo convention `research/NNN-slug.md`, not the `docs/research/` path in the ticket text, matching 118/119/121/122)
- **[research/136-blind-run-synthesis.md](https://github.com/ChmaraX/forensix/blob/63c88f2/research/136-blind-run-synthesis.md)** — companion meta-brief
- Evidence under [`research/_raw/`](https://github.com/ChmaraX/forensix/tree/63c88f2/research/_raw): fetch manifests, HF tree listings, licence captures, configs, Curlie verification

**Done-when check:** ranked shortlist of 6 spanning 4 approach families (3b static-embedding / 3 contextual-embedding / 2 entailment / 5 lists), all surviving the hard filters, enumeration recorded as reproducible API sweeps (incl. one invalid sweep documented so it isn't reused):

| Rank | Candidate | Family | Total bytes | Licence |
|---|---|---|---|---|
| 1 | `minishlab/potion-base-2M` | 3b | 8,468,748 | MIT |
| 2 | `minishlab/potion-base-8M` | 3b | 31,146,255 | MIT |
| 3 | `TaylorAI/bge-micro-v2` | 3 | 18,356,061 | MIT |
| 4 | `sentence-transformers/all-MiniLM-L6-v2` | 3 | 23,725,590 | Apache-2.0 |
| 5 | `cross-encoder/nli-deberta-v3-xsmall` | 2 (adjudication only) | 98,501,034 | Apache-2.0 |
| 6 | UT1 compiled domain table | 5 (tier 0) | build-time | CC BY-SA 4.0 |

Corrections to #119 landed with primary evidence: bge-small's only upstream ONNX is 133 MB (no quantised variant — the ~25 MB figure was a mirror artifact), and UT1 is CC BY-SA 4.0 copyleft. Proposed 15-label forensic taxonomy in §7, incl. mandatory `Unclassified`.

The synthesis brief adds architecture-level corroboration: two context-isolated runs (one with web access, one without) independently converged on the same list-tier → single-pass → abstain shape. It also verified Curlie as a licence-cleared label source (2.9M entries, TSV, CC BY 3.0, 177,289,960 B, ~6-months-stale snapshot) — consumed by #137, not here.

No benchmark numbers anywhere in either brief, per ticket rule. Measurement → #137, decision → #126.

**Comment by ChmaraX** [comment]
## Spike results: Gap 5 + Gap 6 (early) — potion-base-2M Node.js end-to-end

Full report: `/tmp/potion-node-spike/SPIKE-REPORT.md`

### Verdicts

| Criterion | Result |
|---|---|
| **NODE-ONLY** | ✅ **CONFIRMED** |
| **OFFLINE** | ✅ **CONFIRMED** |
| **DETERMINISTIC-SAME-ARCH** | ✅ **CONFIRMED** |
| **DETERMINISTIC-CROSS-ARCH** | ✅ **CONFIRMED** |

All four pass. Every run produced:
```
EMBEDDING_SHA256=167d670b4369880bcffd7cb7f166a85fd86744a7dc483a617a3b07a01351a387
```

### Stack

```
node:22-slim (Docker)
@huggingface/transformers  3.8.1   ← tokenizer (pure JS, BertTokenizer from local tokenizer.json)
onnxruntime-node           1.21.0  ← ONNX inference (bundled binary, --ignore-scripts works)
```

### `npm ci --ignore-scripts` status

onnxruntime-node has `hasInstallScript: true` in the lockfile. With `--ignore-scripts` the binary still loads because it is **bundled inside the npm tarball** — the postinstall hook is verification-only, not a download. CONFIRMED with an isolated `--network none` test before Phase B.

One dependency-management constraint surfaced: `@huggingface/transformers@3.8.1` nests its own `onnxruntime-node@1.21.0`. If you install a different top-level ORT version (e.g. `^1.20.1` resolves to 1.27.0), both versions share the SONAME `libonnxruntime.so.1` and the one loaded first wins — the other's binding fails with a symbol-version error. **Fix: pin `"onnxruntime-node": "1.21.0"` in `package.json` so npm hoists one shared copy.**

### ONNX interface (Gap 5 structural finding)

Model2Vec uses **EmbeddingBag** semantics — *not* the standard `[batch, seq_len]` transformer layout:

```
ONNX_INPUTS  = input_ids, offsets
ONNX_OUTPUTS = embeddings
```

- `input_ids`: flat 1-D array `[seq_len]` of int64 token IDs
- `offsets`: `[0]` (one bag, starting at position 0) for a single sequence
- Output: `[1, 64]` float32 (64-dim after PCA, L2-normalised, Zipf-weighted — all baked into the graph)

Standard transformer integration code that sends `[batch, seq_len]` + `attention_mask` will fail; the EmbeddingBag layout must be used.

### Cross-arch determinism (Gap 6 early)

arm64 native vs amd64 emulated (QEMU), same artifacts, `--network none`:

```
Run 1  linux/arm64  sha256: 167d670b...  wall: 78 ms
Run 2  linux/arm64  sha256: 167d670b...  wall: 70 ms   ← same-arch check
Run 3  linux/x64    sha256: 167d670b...  wall: 824 ms  ← cross-arch (QEMU overhead)
```

Bitwise-identical, as the survey brief predicted: inference is integer table-lookup + floating-point mean over a fixed token order — no attention kernels, no SIMD reduction ordering, no int8 dequantisation path divergence. The cross-arch non-determinism concern from §4 of the brief applies to the contextual transformer ranks (3–5), not to this family-3b model.

### Artifact SHAs (authoritative repo `minishlab/potion-base-2M`)

```
onnx/model.onnx          7,563,349 B  sha256: 92d4a757...
tokenizer.json             683,666 B  sha256: e67e803f...
tokenizer_config.json        1,431 B  sha256: 6725995e...
config.json                    200 B  sha256: b2a89173...
special_tokens_map.json        134 B  sha256: a9e8fb6f...
```

### What this does not answer

- Accuracy of cosine-similarity matching against the 15-label taxonomy (harness ticket)
- Warm-session throughput at 10⁴–10⁵ row scale (78 ms is cold-start incl. tokenizer + ORT session init)
- Gap 6 proper: cross-arch determinism for quantised contextual ranks 3–5 (needs a separate spike)
